docs/design/jet/025-jet-job-member-pruning.md

```diff
@@ -0,0 +1,193 @@
+behavior and applicable to DAGs from any source (SQL/Jet), whereas the _SQL-oriented_
+approach focuses on hinting `ExecutionPlanBuilder` with additional meta-information from
+JobConfig and DAG constructed by SQL optimizer.
+
+##### Generic approach
+
+The core idea here is that `ExecutionPlanBuilder` would have detailed analysis of the received DAG and possibly 
+try to optimize it before creating execution plans. The requirements to prune a member are:
+- each vertex in the DAG may work without an input;
+- DAG does not contain `distributed-broadcast` edges;
+
+This approach generifies member pruning for SQL, Pipeline API and any connector.
+However, it brings high complexity in problem analysis, long development time
+and big list of corner cases, which are discussed in the chapter below.
+
+##### SQL-oriented approach
+
+Unlike the way described above, we can think that member pruning would be beneficial mostly for small jobs.
+The overwhelming majority of such jobs are generated by SQL, and our team effort is to replace Predicate API,
+where `PartitionPredicate` is available for partitioned data querying.
+For that, we can use SQL optimization phase also for determining partitions and for all relations, we can extract information
+about required partitions to run to `JobConfig`, and then mark DAG as 'ableToPruneMembers'. This mark is an indicator
+for `ExecutionPlanBuilder` to choose code path with member pruning during execution plan creation.
+
+### Processor pruning
+
+For processor pruning we would like to separate two kinds of pruning: **intra-member** processor pruning, which
+is applicable only for processor creation elimination inside one member, and **inter-member** processor pruning.
+
+#### Use cases for processor pruning
+
+##### Processor pruning for Scan -> Transform -> Aggregate
+
+1. Single-staged aggregation.
+
+![Scan + Scan -> HashJoin](https://svgshare.com/i/tTM.svg)
+
+On the picture above we can see optimized example from member pruning case. We can go further and just don't create 
+processors which are not participating in data processing by tuning local parallelism parameter with 
+partition involvement knowledge: 
+
+![Scan + Scan -> HashJoin](https://svgshare.com/i/tUD.svg)
+
+So, it's a good target for **intra-member** processor pruning.
+
+2. Double-staged aggregation.
+```
+- COMBINE
+  - ACCUMULATE
+    - TRANSFORM
+      - SCAN[map, partitionKey=K]
+```
+![Scan + Scan -> HashJoin](https://jet-start.sh/docs/assets/arch-dag-4.svg)
+
+For 2-stage aggregation, honestly, we don't see any processor pruning possibilities. Potentially,
+it may be done in another way: we can translate 2-staged into single-staged on SQL opt phase and
+then eliminate member(s) via member pruning, namely scan and flatmap nodes.
+In practice, processor logic is opaque to Jet and may have side effects which may break the job. 
+This example illustrates **inter-member** processor pruning.
```

Comment:
```suggestion
Two-stage aggregation might be a good target for **inter-member** processor pruning. However,
currently, processor logic is opaque to Jet and changing it requires a lot of changes, which may
jeopardize the correctness of execution plans. Potentially, it may be done in another way: we can
translate two-staged into single-staged on SQL opt phase and then eliminate member(s) via member
pruning, namely scan and flatmap nodes.
```

Reply:
> currently, processor logic is opaque to Jet 

It was opaque forever, so this phrasing is incorrect. I'd preserve original text.

Reply:
1. **Currently** doesn't necessarily imply that something has changed.
2. **In practice** implies that what is in practice differs from the theory.
3. The example doesn't illustrate inter-process pruning, but our worries about implementing it.
4. The clause **may have side effects which may break the job** refers to **processor logic**, so it means "processor logic may have side effects which may break the job," which is nonsense.

Thus, I think the original paragraph needs to be changed and the one I propose is a good candidate.

Reply:
I removed the word 'currently.'